carma_cooperative_perception/src/msg_conversion.cpp

```diff
@@ -265,4 +382,150 @@ auto to_external_object_list_msg(
+    sdsm_detected_object.detected_object_common_data.pos =
+      calc_relative_position(current_pose, sdsm_detected_object.detected_object_common_data.pos);
+
+    detected_object_list.detected_object_data.push_back(sdsm_detected_object);
+  }
+  sdsm.objects = detected_object_list;
+
+  return sdsm;
+}
+
+auto to_detected_object_data_msg(
+  const carma_perception_msgs::msg::ExternalObject & external_object,
+ const std::shared_ptr<lanelet::projection::LocalFrameProjector>& map_projection) noexcept
+  -> carma_v2x_msgs::msg::DetectedObjectData
+{
+  carma_v2x_msgs::msg::DetectedObjectData detected_object_data;
+  detected_object_data.presence_vector = 0;
+
+  carma_v2x_msgs::msg::DetectedObjectCommonData detected_object_common_data;
+  detected_object_common_data.presence_vector = 0;
+
+  // common data //////////
+
+  // obj_type_conf - convert from percentile, cast to proper uint type
+  if (external_object.presence_vector & external_object.OBJECT_TYPE_PRESENCE_VECTOR){
+    detected_object_common_data.obj_type_cfd.classification_confidence = static_cast<std::uint8_t>(external_object.confidence * 100);
+  }
+  
+  // detected_id - cast proper type
+  if (external_object.presence_vector & external_object.ID_PRESENCE_VECTOR){
+    detected_object_common_data.detected_id.object_id = static_cast<std::uint16_t>(external_object.id);
+  }
+
+  // pos - Add offset to ref_pos to get object position in map frame -> convert to WGS84 coordinates for sdsm
+  // To get offset: Subtract the external object pose from the current vehicle location given by the current_pose topic
+  if (external_object.presence_vector & external_object.POSE_PRESENCE_VECTOR){
+    detected_object_common_data.pos.offset_x.object_distance = static_cast<float>(external_object.pose.pose.position.x);
+    detected_object_common_data.pos.offset_y.object_distance = static_cast<float>(external_object.pose.pose.position.y);
+    detected_object_common_data.pos.offset_z.object_distance = static_cast<float>(external_object.pose.pose.position.z);
+  }
+
+  // speed/speed_z - convert vector velocity to scalar speed val given x/y components
+  if (external_object.presence_vector & external_object.VELOCITY_INST_PRESENCE_VECTOR){
+    detected_object_common_data.speed.speed = std::hypot(external_object.velocity_inst.twist.linear.x, external_object.velocity_inst.twist.linear.y);
+
+    detected_object_common_data.presence_vector |= carma_v2x_msgs::msg::DetectedObjectCommonData::HAS_SPEED_Z;
+    detected_object_common_data.speed_z.speed = external_object.velocity_inst.twist.linear.z;
+
+    // TODO: heading - convert ang vel to scale heading
+    lanelet::BasicPoint3d external_object_position{external_object.pose.pose.position.x, external_object.pose.pose.position.y,external_object.pose.pose.position.z};
+    detected_object_common_data.heading.heading = remove_units(enu_orientation_to_true_heading(external_object.velocity_inst.twist.angular.z, external_object_position, map_projection));
```

Comment:
This calculation looks incorrect. The `enu_orientation_to_true_heading` function expects a yaw angle, but the argument passed in is an agular velocity.